### PR TITLE
Feature/holo 808 create holograph drop

### PR DESCRIPTION
**Pull request by alexanderattar**
## Describe Changes

- Add new logic to allow deploying a drop contract
- Fixes issue with the network monitor not properly keeping track of the nonce
- Update tokenUriType to uriType and tokenUri to just uri to support other instances where the uri isn't necessarily token, but is a drop / edition
- 
image

Example saved file:

```
{
  "version": "beta3",
  "deploymentConfig": {
    "config": {
      "contractType": "0x00000000000000000000000000486f6c6f677261706845524337323144726f70",
      "chainType": "0xee6b2803",
      "salt": "0x000000000000000000000000000000000000000000000000000000000095825a",
      "byteCode": "0x608060405234801561001057600080fd5b50610897806100206000396000f3fe6080604052600436106100695760003560e01c80634ddf47d4116100435780634ddf47d41461017e5780638da5cb5b146101cf5780638f32d59b14610209576100d5565b806301ffc9a7146101135780632f54bf6e146101495780633ccfd60b14610169576100d5565b366100d5576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601760248201527f44524f50533a20756e726561636861626c6520636f646500000000000000000060448201526064015b60405180910390fd5b337fe9fcff60011c1a99f7b7244d1f2d9da93d79ea8ef3654ce590d775575255b2bd54146001811461010657600080fd5b600160805260206080f35b005b34801561011f57600080fd5b5061013461012e36600461070d565b50600090565b60405190151581526020015b60405180910390f35b34801561015557600080fd5b50610134610164366004610756565b61021e565b34801561017557600080fd5b5061011161027d565b34801561018a57600080fd5b5061019e6101993660046107bb565b6104b7565b6040517fffffffff000000000000000000000000000000000000000000000000000000009091168152602001610140565b3480156101db57600080fd5b506101e46105e2565b60405173ffffffffffffffffffffffffffffffffffffffff9091168152602001610140565b34801561021557600080fd5b50610134610611565b60006102487fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775490565b73ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff16149050919050565b7fe9fcff60011c1a99f7b7244d1f2d9da93d79ea8ef3654ce590d775575255b2bd5473ffffffffffffffffffffffffffffffffffffffff163303610399577fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775473ffffffffffffffffffffffffffffffffffffffff16367fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0013573ffffffffffffffffffffffffffffffffffffffff1614610394576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601a60248201527f44524f50533a206f776e6572206f6e6c792066756e6374696f6e00000000000060448201526064016100cc565b61044f565b7fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775473ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161461044f576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601a60248201527f44524f50533a206f776e6572206f6e6c792066756e6374696f6e00000000000060448201526064016100cc565b7fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775460405173ffffffffffffffffffffffffffffffffffffffff909116904780156108fc02916000818181858888f193505050501580156104b4573d6000803e3d6000fd5b50565b60006104e17f4e5f991bca30eca2d4643aaefa807e88f96a4a97398933d572a3c0d973004a015490565b15610548576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601a60248201527f44524f50533a20616c726561647920696e697469616c697a656400000000000060448201526064016100cc565b337fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727778190557fe9fcff60011c1a99f7b7244d1f2d9da93d79ea8ef3654ce590d775575255b2bd8190556105ba60017f4e5f991bca30eca2d4643aaefa807e88f96a4a97398933d572a3c0d973004a0155565b507f4ddf47d40000000000000000000000000000000000000000000000000000000092915050565b600061060c7fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775490565b905090565b600061063b7fe9fcff60011c1a99f7b7244d1f2d9da93d79ea8ef3654ce590d775575255b2bd5490565b73ffffffffffffffffffffffffffffffffffffffff1633036106cf577fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775473ffffffffffffffffffffffffffffffffffffffff16367fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0013573ffffffffffffffffffffffffffffffffffffffff1614905090565b507fb56711ba6bd3ded7639fc335ee7524fe668a79d7558c85992e3f8494cf7727775473ffffffffffffffffffffffffffffffffffffffff16331490565b60006020828403121561071f57600080fd5b81357fffffffff000000000000000000000000000000000000000000000000000000008116811461074f57600080fd5b9392505050565b60006020828403121561076857600080fd5b813573ffffffffffffffffffffffffffffffffffffffff8116811461074f57600080fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6000602082840312156107cd57600080fd5b813567ffffffffffffffff808211156107e557600080fd5b818401915084601f8301126107f957600080fd5b81358181111561080b5761080b61078c565b604051601f82017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0908116603f011681019083821181831017156108515761085161078c565b8160405282815287602084870101111561086a57600080fd5b82602086016020830137600092810160200192909252509594505050505056fea164736f6c634300080d000a",
      "initCode": "0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000aaeb6d7670e522a718067333cd4e000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c0000000000000000000000000d078e391cbaeaa6c5785124a7207ff57d64604b7000000000000000000000000d078e391cbaeaa6c5785124a7207ff57d64604b7000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000064000000000000000000000000000000000000000000000000000000000000020000000000000000000000000050c9a9e0b12d47531cce7ece657ec1f1228aa4b90000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000000461736466000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004415344460000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000046173646600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000b697066733a2f2f73616466000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c616e696d6174696f6e5552490000000000000000000000000000000000000000"
    },
    "signature": {
      "r": "0x3a7c8dcc2760b238a327ad752814124d7d0c79bb1afe261ea147e8129fc2b88b",
      "s": "0x1ceedbddf4921d7b64401db0912e0e7f03afa60837c8b6cb71254846771dc149",
      "v": 27
    },
    "signer": "0xd078E391cBAEAa6C5785124a7207ff57d64604b7"
  },
  "metadata": {
    "collectionName": "asdf",
    "collectionSymbol": "ASDF",
    "royaltyBps": 100,
    "contractType": "HolographERC721Drop",
    "description": "asdf",
    "imageURI": "ipfs://sadf",
    "numOfEditions": 10
  },
  "transactions": [
    {
      "address": "0x3f56f2085981156e9987038b2ab44e8dfeb22560",
      "txHash": "0xa2d71cc2f8d660d42effdc997ecf6bb01e255a7a2563a6d6249fe0b2bc75ad22",
      "blockNumber": 8609126,
      "network": "ethereumTestnetGoerli"
    },
    {
      "address": "0x3f56f2085981156e9987038b2ab44e8dfeb22560",
      "txHash": "0xcc366be9cbf756c8aae4b44eb8ac732a378fabcb382fbd9f2bff19a8ab62faed",
      "blockNumber": 19592351,
      "network": "avalancheTestnet"
    }
  ]
}
```

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] I have checked eslint

### CODE REVIEW

File: src/commands/create/contract.ts [hunk 1]
```diff
@@ -170,14 +182,28 @@ export default class Contract extends Command {
...
+
         deploymentConfig.config.salt = salt
         bytecodeType = await checkBytecodeTypeFlag(undefined, 'Select the bytecode type to deploy')
-        contractType =
-          bytecodeType === BytecodeType.Custom
-            ? await checkOptionFlag(contractTypes, undefined, 'Select the contract type to create')
-            : bytecodeType === BytecodeType.SampleERC20
-            ? 'HolographERC20'
-            : 'HolographERC721'
+
```
Comment (alexanderattar) [inline src/commands/create/contract.ts:197]: prefer readability over cleverness

File: README.md [hunk 2]
```diff
@@ -355,15 +355,15 @@ FLAGS
...
+  --uri=<value>           The uri of the token, minus the prepend (ie "ipfs://")
+  --uriType=<option>      The token URI type
                                <options: ipfs|https|arweave>
 
 DESCRIPTION
   Mint a Holographable NFT.
 
 EXAMPLES
-  $ holograph create:nft --network="goerli" --collectionAddress="0xf90c33d5ef88a9d84d4d61f62c913ba192091fe7" --tokenId="0" --tokenUriType="ipfs" --tokenUri="QmfQhPGMAbHL31qcqAEYpSP5gXwXWQa3HZjkNVzZ2mRsRs/metadata.json" --env mainnet|testnet|develop|experimental
+  $ holograph create:nft --network="goerli" --collectionAddress="0xf90c33d5ef88a9d84d4d61f62c913ba192091fe7" --tokenId="0" --uriType="ipfs" --uri="QmfQhPGMAbHL31qcqAEYpSP5gXwXWQa3HZjkNVzZ2mRsRs/metadata.json" --env mainnet|testnet|develop|experimental
```
Comment (sogoiii) [inline README.md:366]: @alexanderattar so why did you change the name? I think we can change it to be clearer. I just want to make sure people don't confuse the image CID and the metadata CID. 

Also if we make this change, we have to re-generate the readme file.

File: src/commands/create/contract.ts [hunk 3]
```diff
@@ -394,24 +490,29 @@ export default class Contract extends Command {
...
-      throw new Error('failed to confirm that the transaction was mined')
+      throw new Error('Failed to confirm that the transaction was mined')
     } else {
       const logs: any[] | undefined = this.networkMonitor.decodeBridgeableContractDeployedEvent(
         receipt,
         this.networkMonitor.factoryAddress,
       )
       if (logs === undefined) {
-        throw new Error('failed to extract transfer event from transaction receipt')
+        throw new Error('Failed to extract transfer event from transaction receipt')
```
Comment (sogoiii) [inline src/commands/create/contract.ts:592]: nice catch.

**Review by sogoiii — CHANGES_REQUESTED** [review]
I tried to deploy, and i got a bunch of these in my terminal.

```
sendTransaction{
  "reason": "processing response error",
  "code": "SERVER_ERROR",
  "body": "{\"jsonrpc\":\"2.0\",\"id\":349,\"error\":{\"code\":-32000,\"message\":\"transaction underpriced: address 0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0 have gas fee cap (0) \\u003c pool minimum fee cap (25000000000)\"}}\n",
  "error": {
    "code": -32000
  },
  "requestBody": "{\"method\":\"eth_sendRawTransaction\",\"params\":[SOME_LARGE_BYTECODE]
```
It did eventually work, but i dont know why it failed like 5 times before it worked.


Also, I have a feature request. When the deployment file is saved, it does not include the resulting contract address. 

```
{
  "config": {
    "contractType": "0x00000000000000000000000000486f6c6f677261706845524337323144726f70",
    "chainType": "0xee6b2803",
    "salt": "0x00000000000000000000000000000000000000000000000000000000000004d2",
    "byteCode": "SOME_LARGE_BYTES",
    "initCode": "SOME_LARGE_BYTES"
  },
  "signature": {
    "r": "0xfe370cc9d7102ed045dea2409e9d39ef031f5be151b8705c0a98a675200b56f4",
    "s": "0x497a8bca118bc1a462dca336dcded2e4f70f816ae89d2e367828d1e5226859fc",
    "v": 27
  },
  "signer": "0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0"
}
```

It would be great that the contract address be in here so its easier to know what got deployed. It should also have some of the input field information.

```
{
  "collection": {
     "address": "0xABC",
     "contractType": "HolographERC721Drop",
     "name" : "NAME",
     "symbol": "SYMBOL",
     "nft": {
         "uriType": "ipfs",
         "uri": "URICID",
     },
     "drops": {
       "type": "EDITIONS",
       "supply": "500",
      }
  },
  "transactions": [{"chainId": "CHAINID", "tx": "0xSOMETX"}],
  ... //the original stuff from before
}
```

Open to other key names, as long as it make sense.

**Comment by sogoiii** [comment]
@alexanderattar oh i forgot to mention, that we are going out with open Editions as the drop type. So we don't have to set a number of editions. The supply (number) is only necessary for limited editions.

File: README.md (hunk 2 above)
Comment (alexanderattar) [reply README.md:366]: I updated the CLI code to reflect these changes so I believe if we regenerate the readme it should stay the same, but we can try that out

Comment (alexanderattar) [reply README.md:366]: I changed the name to acknowledge that this flag can be used to send and validate not just tokenUris but other uses such as the drop metadata uri. We can duplicate the logic to keep things more specific if that's preferred. lmk what you think

**Comment by alexanderattar** [comment]
> @alexanderattar oh i forgot to mention, that we are going out with open Editions as the drop type. So we don't have to set a number of editions. The supply (number) is only necessary for limited editions.

yes that's right. in this case we will just select zero to toggle it to open editions mode

**Comment by alexanderattar** [comment]
> I tried to deploy, and i got a bunch of these in my terminal.
> 
> ```
> sendTransaction{
>   "reason": "processing response error",
>   "code": "SERVER_ERROR",
>   "body": "{\"jsonrpc\":\"2.0\",\"id\":349,\"error\":{\"code\":-32000,\"message\":\"transaction underpriced: address 0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0 have gas fee cap (0) \\u003c pool minimum fee cap (25000000000)\"}}\n",
>   "error": {
>     "code": -32000
>   },
>   "requestBody": "{\"method\":\"eth_sendRawTransaction\",\"params\":[SOME_LARGE_BYTECODE]
> ```
> 
> It did eventually work, but i dont know why it failed like 5 times before it worked.
> 
> Also, I have a feature request. When the deployment file is saved, it does not include the resulting contract address.
> 
> ```
> {
>   "config": {
>     "contractType": "0x00000000000000000000000000486f6c6f677261706845524337323144726f70",
>     "chainType": "0xee6b2803",
>     "salt": "0x00000000000000000000000000000000000000000000000000000000000004d2",
>     "byteCode": "SOME_LARGE_BYTES",
>     "initCode": "SOME_LARGE_BYTES"
>   },
>   "signature": {
>     "r": "0xfe370cc9d7102ed045dea2409e9d39ef031f5be151b8705c0a98a675200b56f4",
>     "s": "0x497a8bca118bc1a462dca336dcded2e4f70f816ae89d2e367828d1e5226859fc",
>     "v": 27
>   },
>   "signer": "0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0"
> }
> ```
> 
> It would be great that the contract address be in here so its easier to know what got deployed. It should also have some of the input field information.
> 
> ```
> {
>   "collection": {
>      "address": "0xABC",
>      "contractType": "HolographERC721Drop",
>      "name" : "NAME",
>      "symbol": "SYMBOL",
>      "nft": {
>          "uriType": "ipfs",
>          "uri": "URICID",
>      },
>      "drops": {
>        "type": "EDITIONS",
>        "supply": "500",
>       }
>   },
>   "transactions": [{"chainId": "CHAINID", "tx": "0xSOMETX"}],
>   ... //the original stuff from before
> }
> ```
> 
> Open to other key names, as long as it make sense.

the transaction underpriced thing has to do with the gas service calculating a reasonable gas price before submitting it so it always errors like that at first as far as i've seen from whatever it's doing to gather network condition data. i'm not a huge fan of that stuff as it was difficult to try to understand and debug, but is does eventually work and saves users gas so i don't want to mess with it too much more at the moment since it's in the network monitor and will effect operators

**Comment by alexanderattar** [comment]
the feature request is reasonable. i think i should be able to add those keys relatively easily when the file is saved. i'll work on that next

**Review by BravoNatalie — APPROVED** [review]

**Review by sogoiii — CHANGES_REQUESTED** [review]
## Steps Copy Change

For the step that asks for the URI, we should make clear that we are asking for the imageURI. I feel people will probably make this mistake. 

We can probably remove the step that asks for the number of editions. We don't support capped editions at all yet. 


## Start and end times

We are doing drops that have a limited time availability. We need to have a way to set the start  and end times. 


## Possible Bug

I was able to deploy this drop onto mumbai on first attempt. When I try to deploy by using an existing deployment config, onto either fuji or goerli, the command fails. 

```
➜ ./bin/dev create:contract --env develop
Loading user configurations...
? Please enter the password to decrypt the private key for 0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0 [hidden]

👉 Environment: develop

User configurations loaded.
Loading network RPC providers... done
? Select the type of deployment to use existing deployment config
? Enter the config file to use dropsample2
? Select the network on which the contract will be executed goerli
Checking that contract is not already deployed on "ethereumTestnetGoerli" network... done
? Next steps submit the transaction, would you like to proceed? Yes
[2023-03-07T01:43:30.974Z] [Contract] [Ethereum Goerli] [1197f943] Executing contract function deployHolographableContract
[2023-03-07T01:43:30.976Z] [Contract] [Ethereum Goerli] [1197f943] Gas price is 3.094842883 GWEI
[2023-03-07T01:43:31.262Z] [Contract] [Ethereum Goerli] [1197f943] {"reason":"execution reverted","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x4BB282b41b17cC0EDD8cED6eC7702Dba039b4AB0","gasPrice":{"type":"BigNumber","hex":"0x0170ef1c06"},"to":"0x90425798cc0e33932f11edc3EeDBD4f3f88DFF64","value":{"type":"BigNumber","hex":"0x00"},"data":"BIGBLOB","accessList":null},"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":60,\"error\":{\"code\":-32000,\"message\":\"execution reverted\"}}","error":{"code":-32000},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"gasPrice\":\"0x170ef1c06\",\"value\":\"0x0\",\"from\":\"0x4bb282b41b17cc0edd8ced6ec7702dba039b4ab0\",\"to\":\"0x90425798cc0e33932f11edc3eedbd4f3f88dff64\",\"data\":\"BIGBLOB\"}],

```

**Comment by alexanderattar** [comment]
Hmm I was able to deploy on Goerli and then use the file to deploy on Fuji without issue. I'll look into this. Might be a Mumbai thing again

**Review by sogoiii — APPROVED** [review]
👍 I see the questions were added. Nice.
